pkg/controller/ocmagent/ocmagent_controller.go

```diff
@@ -0,0 +1,215 @@
+package ocmagent
+
+import (
+	"context"
+	"github.com/go-logr/logr"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	ocmagentv1alpha1 "github.com/openshift/ocm-agent-operator/pkg/apis/ocmagent/v1alpha1"
+	ctrlconst "github.com/openshift/ocm-agent-operator/pkg/consts/controller"
+	oahconst "github.com/openshift/ocm-agent-operator/pkg/consts/ocmagenthandler"
+	"github.com/openshift/ocm-agent-operator/pkg/ocmagenthandler"
+)
+
+// ReconcileOCMAgent reconciles a OCMAgent object
+type ReconcileOCMAgent struct {
+	Client client.Client
+	Scheme *runtime.Scheme
+	Ctx    context.Context
+	Log    logr.Logger
+
+	OCMAgentHandler ocmagenthandler.OCMAgentHandler
+}
+
+var log = logf.Log.WithName("controller_ocmagent")
+
+// Add creates a new OCMAgent Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new ReconcileOCMAgent
+func newReconciler(mgr manager.Manager) *ReconcileOCMAgent {
+	client := mgr.GetClient()
+	log := ctrl.Log.WithName("controllers").WithName("OCMAgent")
+	ctx := context.Background()
+	scheme := mgr.GetScheme()
+	return &ReconcileOCMAgent{
+		Client:          client,
+		Scheme:          scheme,
+		Ctx:             ctx,
+		Log:             log,
+		OCMAgentHandler: ocmagenthandler.New(client, scheme, log, ctx),
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r *ReconcileOCMAgent) error {
+	// Create a new controller
+	c, err := controller.New("ocmagent-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource OCMAgent
+	err = c.Watch(&source.Kind{Type: &ocmagentv1alpha1.OcmAgent{}}, &handler.EnqueueRequestForObject{})
```

Comment:
as we discussed in sync meet, not sure if we can watch the resource in the specified ns only to improve the performance?

Reply:
Yes, based on the last meet I updated the `Deployment` to refine `WATCH_NAMESPACE` to watch both the `openshift-ocm-agent-operator` and `openshift-config` namespaces which should vastly cut down on the number of reconcile requests it receives
:
https://github.com/openshift/ocm-agent-operator/blob/5fcc9b0d1c2024e3a462e7283e131fd12882d0dc/deploy/50_ocm-agent-operator.Deployment.yaml#L36